go/cli/mcap/cmd/filter.go

```diff
@@ -0,0 +1,326 @@
+					die("error closing read source: %s", closeErr)
+				}
+			}()
+			reader = newReader
+		}
+
+		var writer io.Writer
+		if filterOutput == "" {
+			if !utils.StdoutRedirected() {
+				die("Binary output can screw up your terminal. Supply -o or redirect to a file or pipe")
+			}
+			writer = os.Stdout
+		} else {
+			newWriter, err := os.Create(filterOutput)
+			if err != nil {
+				die("failed to open %s for writing: %s", filterOutput, err)
+			}
+			defer func() {
+				if err := newWriter.Close(); err != nil {
+					die("error closing write target: %s", err)
+				}
+			}()
+			writer = newWriter
+		}
+
+		err = filter(reader, writer, filterOptions)
+		if err != nil {
+			die(error.Error(err))
+		}
+	},
+}
+
+func compileMatchers(regexStrings []string) ([]regexp.Regexp, error) {
+	var matchers []regexp.Regexp
+
+	for _, regexString := range regexStrings {
+		// auto-surround with ^$ if not specified.
+		if regexString[:1] != "^" {
+			regexString = "^" + regexString
+		}
+		if regexString[len(regexString)-1:] != "$" {
+			regexString = regexString + "$"
+		}
+		regex, err := regexp.Compile(regexString)
+		if err != nil {
+			return nil, err
+		}
+		matchers = append(matchers, *regex)
+	}
+	return matchers, nil
+}
+
+type markableSchema struct {
+	*mcap.Schema
+	written bool
+}
+type markableChannel struct {
+	*mcap.Channel
+	written bool
+}
+
+func filter(
+	r io.Reader,
+	w io.Writer,
+	opts *filterOpts,
+) error {
+	lexer, err := mcap.NewLexer(r)
+	if err != nil {
+		return err
+	}
+	mcapWriter, err := mcap.NewWriter(w, &mcap.WriterOptions{
+		Compression: opts.compressionFormat,
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = mcapWriter.Close()
+	}()
+
+	buf := make([]byte, 1024)
+	schemas := make(map[uint16]markableSchema)
+	channels := make(map[uint16]markableChannel)
+
+	for {
+		token, data, err := lexer.Next(buf)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+		}
+		if len(data) > len(buf) {
+			buf = data
+		}
+		switch token {
+		case mcap.TokenHeader:
+			header, err := mcap.ParseHeader(data)
+			if err != nil {
+				return err
+			}
+			if err = mcapWriter.WriteHeader(header); err != nil {
+				return err
+			}
+		case mcap.TokenSchema:
+			schema, err := mcap.ParseSchema(data)
+			if err != nil {
+				return err
+			}
+			schemas[schema.ID] = markableSchema{schema, false}
+		case mcap.TokenChannel:
+			channel, err := mcap.ParseChannel(data)
+			if err != nil {
+				return err
+			}
+			if len(opts.includeTopics) > 0 {
+				for _, matcher := range opts.includeTopics {
+					if matcher.MatchString(channel.Topic) {
+						channels[channel.ID] = markableChannel{channel, false}
+					}
+				}
+			} else if len(opts.excludeTopics) > 0 {
+				for _, matcher := range opts.excludeTopics {
+					if !matcher.MatchString(channel.Topic) {
+						channels[channel.ID] = markableChannel{channel, false}
+					}
+				}
+			} else {
+				channels[channel.ID] = markableChannel{channel, false}
+			}
```

Comment:
if we write the schema+channel after observing the channel, rather than after observing the first message, would that let us drop the "markable" aspect of the channel map, or maybe even drop the channel map altogether?

That would also result in us returning channels with matching topics even if they have no messages on them, which seems more correct on balance to me. What do you think?

Reply:
well, a channel in an mcap with no messages is definitely something you can record - so why wouldn't the filter command preserve channels on the supplied topic?